packages/itwin/grouping-mapping-widget/src/components/Properties/GroupProperties/GroupPropertyAction.tsx

```diff
@@ -73,121 +73,121 @@ export const GroupPropertyAction = ({
+  const queryClient = useQueryClient();
 
   const reset = useCallback(() => {
     setPropertyName("");
     setDataType(DataType.Undefined);
     setSelectedProperties([]);
   }, []);
 
-  useEffect(() => {
-    const generateProperties = async () => {
-      setIsLoading(true);
+  const fetchProperties = useCallback(async () => {
+    if (!iModelConnection) return;
 
-      if (!iModelConnection) return;
+    const result = await manufactureKeys(group.query, iModelConnection);
+    const descriptor = await fetchPresentationDescriptor(iModelConnection, result);
 
-      const result = await manufactureKeys(group.query, iModelConnection);
+    // Only allow primitives and structs
+    const propertyFields = descriptor?.fields.filter(
+      (field) =>
+        field.type.valueFormat === PropertyValueFormat.Primitive ||
+        field.type.valueFormat === PropertyValueFormat.Struct
+    ) ?? [];
 
-      const descriptor = await fetchPresentationDescriptor(iModelConnection, result);
+    const propertiesMetaData = convertPresentationFields(propertyFields);
 
-      // Only allow primitives and structs
-      const propertyFields =
-        descriptor?.fields.filter(
-          (field) =>
-            field.type.valueFormat === PropertyValueFormat.Primitive ||
-            field.type.valueFormat === PropertyValueFormat.Struct
-        ) ?? [];
-
-      const propertiesMetaData = convertPresentationFields(propertyFields);
+    let groupPropertyDetails = null;
+    if (groupProperty) {
+      const accessToken = await getAccessToken();
+      groupPropertyDetails = await mappingClient.getGroupProperty(
+        accessToken,
+        iModelId,
+        mappingId,
+        group.id,
+        groupProperty.id
+      );
+    }
 
-      setPropertiesMetaData(propertiesMetaData);
+    return { propertiesMetaData, groupPropertyDetails };
+  }, [getAccessToken, group.id, group.query, groupProperty, iModelConnection, iModelId, mappingClient, mappingId]);
 
-      if (groupProperty) {
-        const accessToken = await getAccessToken();
-        let response: GroupProperty | undefined;
-        try {
-          response = await mappingClient.getGroupProperty(
-            accessToken,
-            iModelId,
-            mappingId,
-            group.id,
-            groupProperty.id
-          );
+  const onSuccess = useCallback((data) => {
+    setPropertiesMetaData(data.propertiesMetaData);
 
-          setPropertyName(response.propertyName);
-          setOldPropertyName(response.propertyName);
-          setDataType(response.dataType);
-          setQuantityType(response.quantityType);
-          const properties = findProperties(response.ecProperties, propertiesMetaData);
-          if (properties.length === 0) {
-            setPropertiesNotFoundAlert(true);
-          }
+    if (data.groupPropertyDetails) {
+      const { propertyName, dataType, quantityType, ecProperties } = data.groupPropertyDetails;
```

Comment:
I don't think it's worth deconstructing an object here. Are the individual properties of the object used in more than 1 place?

Reply:
Nope. Changed. Also cleaned it up further. I was using a deprecated useQuery callback.